message_ix/core.py

```diff
@@ -725,56 +725,85 @@ def add_macro(
+            them is retained.
+
+        Raises
+        ------
+        KeyError
+            if `name` is the name of a different kind of item (for instance, a
+            parameter) or of no known item.
+        ValueError
+            if `name` is a set indexed by 1 or more others.
         """
-        try:
-            self.check_out()
-            commit = True
-        except RuntimeError:
-            commit = False
-        keys = list(mapping.keys())
-
-        # search for from_tech in sets and replace
-        for item in self.set_list():
-            ix_set = self.set(item)
-            if isinstance(ix_set, pd.DataFrame):
-                if name in ix_set.columns and not ix_set.empty:
-                    for key, value in mapping.items():
-                        df = ix_set[ix_set[name] == key]
-                        if not df.empty:
-                            df[name] = value
-                            self.add_set(item, df)
-            elif ix_set.isin(keys).any():  # ix_set is pd.Series
-                for key, value in mapping.items():
-                    if ix_set.isin([key]).any():
-                        self.add_set(item, value)
-
-        # search for from_tech in pars and replace
-        for item in self.par_list():
-            if name not in self.idx_names(item):
-                continue
-            for key, value in mapping.items():
-                df = self.par(item, filters={name: [key]})
-                if not df.empty:
-                    df[name] = value
-                    self.add_par(item, df)
-
-        # this removes all instances of from_tech in the model
+        commit = maybe_check_out(self)
+
+        # - Iterate over tuples of (item_name, ix_type); only those indexed by `name`.
+        # - First the set itself; then all other sets; then all parameters.
+        kw: Dict[str, Any] = dict(indexed_by=name, par_data=False)
+        for item_name, ix_type in chain(
+            [(name, "set")],
+            zip_longest(self.items(ItemType.SET, **kw), [], fillvalue="set"),
+            zip_longest(self.items(ItemType.PAR, **kw), [], fillvalue="par"),
+        ):
+            # Identify some index names of this set; only those where the corresponding
+            # index set is `name`
+            names = [
+                n
+                for (s, n) in zip(self.idx_sets(item_name), self.idx_names(item_name))
+                if s == name
+            ]
+
+            # Construct filters for add_{par,set} and an argument for pd.*.replace()
+            if len(names) == 0:
+                # This is the set `name` itself → data will be pd.Series
+                filters = None
+                to_replace: Mapping[str, Union[str, Mapping]] = mapping
+            else:
+                # An indexed set or parameter → data will be pd.DataFrame
+                # Filters: only the values to be replaced
+                filters = {n: mapping.keys() for n in names}
```

Comment:
@khaeru this line creates a mapping of filters for all `names` at the same time. But this should be per each name in `names`. For example, if the code is used to rename "node", this filter will be `{"node_loc": "Westeros", "node_origin": "Westeros"}` for parameter `input`, i.e., the union of both "node_loc" and "node_origin". But the intention should be to retrieve and replace any entry in parameter `input` which has "node_loc" **or** "node_origin" of "Westeros". In the Westeros example as it is, we don't have such a case and as such the tests pass. But in the global model, we have many entries in `input` with "node_loc" being a node, e.g., "R11_AFR" and "node_origin" being another node, usually "R11_GLB". The code you suggested would not retrieve/replace these cases, if existed, `{"node_loc": "Westeros", "node_origin": <another node>}`. I hope I'm not missing something out.

Reply:
Thanks for checking carefully! I believe you're correct.

- The built-in `filters=` functionality of ixmp indeed only provides intersection/AND logic, not union/OR.
- We can extend the test cases here by adding 1+ rows that would meet an OR but not an AND condition, and then confirm they are modified as expected. This test will fail at the moment.
- I will next think of a fix. I think either we'll need to reintroduce a second loop (call .par() once with filters for each named dimension) or retrieve all data for the parameter and then pass it through something like `DataFrame.query()`. Ideally these would perform equally well, but in practice I think they might not.

Reply:
Please see the added test and let me know if the code now appears to cover the case you describe.

Reply:
Thanks @khaeru for reacting quickly. The updated code resolves the issue I mentioned. The new test, `test_rename2`, captures the case very nicely.